#### Project Assignment: Phase 2 - Data Modeling

#### Author: Vidya Prabhu

##### **Instructions:**
##### 1. Extracting Data from MongoDB
##### Python program will retrieve raw data from MongoDB and transform it into a well-structured format suitable for SQLite.
##### connect to either MongoDB Atlas (cloud) or a local MongoDB instance.
##### extract only the data fields relevant to your analysis, ensuring unnecessary information is excluded.
##### 2. Insert Data into SQLite
##### Create an SQLite database with properly structured tables to store the extracted data.
##### Validate that all data values have the correct data types before inserting them into SQLite.
##### Implement error-handling mechanisms to manage missing fields, such as records without lobbying activities or income data.
##### After completing the insertion process, print the total number of rows in the created table to verify the data transfer.
##### This phase ensures that the data is clean, structured, and ready for analysis in SQLite.

##### Step 1: Connecting to mongo db

In [538]:
from pymongo import MongoClient

# Connecting to MongoDB 
client = MongoClient("mongodb+srv://vidyaprabhu96:Myuniverse20@cluster0.6x4bn.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0") 
db = client["ElectricVehiclesDB"]  # Creating database name
collection = db["ElectricVehiclesDataCollections"]  # Creating collection name

##### Step 2: Verfying the connection

##### 1. Verifying the database connection

In [541]:
try:
    # Fetching the list of databases
    db_list = client.list_database_names()

    # Checking if 'ElectricVehiclesDB' exists in the database list
    if "ElectricVehiclesDB" in db_list:
        print("Connection successful! Database exists.") # Success
    else:
        print("Connected, but 'ElectricVehiclesDB' database not found.") # failure
except Exception as e:
    # Handling connection errors and printing the error message
    print("Connection failed:", e)

Connection successful! Database exists.


##### 2. Verifying the collection connection and displaying one record

In [543]:
try:
    # Fetching list of collections in the database
    collection_list = db.list_collection_names()

     # Checking if 'ElectricVehiclesDataCollections' exists in the collection list
    if "ElectricVehiclesDataCollections" in collection_list:
        print("Connection successful! Collection exists.")
        
        # Fetching a sample document from the collection
        sample_doc = collection.find_one()
        if sample_doc:
            print("Sample document:", sample_doc)
        else:
            print("Connected, but no documents found in the collection.") 
    
        # Counting the number of documents in the collection
        doc_count = collection.count_documents({})
        print("Total number of documents:", doc_count) # displaying the total records
    
    else:
        print("Connected, but 'ElectricVehiclesDataCollections' collection not found.")  # Failure

except Exception as e:
    # Handle connection errors and print the error message
    print("Connection failed:", e)

Connection successful! Collection exists.
Sample document: {'_id': ObjectId('67b36f02e23a723cde1e90de'), 'sid': 'row-9exd_xzw7-2hfk', 'id': '00000000-0000-0000-F8E5-5FABDC1A77AA', 'position': 0, 'created_at': 1739484275, 'created_meta': None, 'updated_at': 1739484437, 'updated_meta': None, 'meta': '{ }', 'VIN (1-10)': '2T3YL4DV0E', 'County': 'King', 'City': 'Bellevue', 'State': 'WA', 'Postal Code': '98005', 'Model Year': '2014', 'Make': 'TOYOTA', 'Model': 'RAV4', 'Electric Vehicle Type': 'Battery Electric Vehicle (BEV)', 'Clean Alternative Fuel Vehicle (CAFV) Eligibility': 'Clean Alternative Fuel Vehicle Eligible', 'Electric Range': '103', 'Base MSRP': '0', 'Legislative District': '41', 'DOL Vehicle ID': '186450183', 'Vehicle Location': 'POINT (-122.1621 47.64441)', 'Electric Utility': 'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)', '2020 Census Tract': '53033023604', 'Counties': '3009', 'Congressional Districts': '9', 'WAOFM - GIS - Legislative District Boundary': '49'}
Total number 

##### Step 3: Extracting only the data fields that are relevant to my planned data analysis task

In [545]:
try:
    # Defining the required fields
    required_fields = {
        "VIN (1-10)": 1, "County": 1, "City": 1, "State": 1, "Postal Code": 1,
        "Model Year": 1, "Make": 1, "Model": 1, "Electric Vehicle Type": 1,
        "Clean Alternative Fuel Vehicle (CAFV) Eligibility": 1, "Electric Range": 1,
        "Base MSRP": 1, "Legislative District": 1, "DOL Vehicle ID": 1,
        "Vehicle Location": 1, "Electric Utility": 1, "2020 Census Tract": 1,
        "Counties": 1, "Congressional Districts": 1, "WAOFM - GIS - Legislative District Boundary": 1
    }

    # Fetching only the required fields from MongoDB
    extracted_data = list(collection.find({}, required_fields))

    # Printing the first 5 records as a sample
    for record in extracted_data[:5]:
        print(record)

    print("Total extracted records:", len(extracted_data)) # checking the total count

except Exception as e:
    # Handling errors and printing the error message
    print("Data extraction failed:", e)

{'_id': ObjectId('67b36f02e23a723cde1e90de'), 'VIN (1-10)': '2T3YL4DV0E', 'County': 'King', 'City': 'Bellevue', 'State': 'WA', 'Postal Code': '98005', 'Model Year': '2014', 'Make': 'TOYOTA', 'Model': 'RAV4', 'Electric Vehicle Type': 'Battery Electric Vehicle (BEV)', 'Clean Alternative Fuel Vehicle (CAFV) Eligibility': 'Clean Alternative Fuel Vehicle Eligible', 'Electric Range': '103', 'Base MSRP': '0', 'Legislative District': '41', 'DOL Vehicle ID': '186450183', 'Vehicle Location': 'POINT (-122.1621 47.64441)', 'Electric Utility': 'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)', '2020 Census Tract': '53033023604', 'Counties': '3009', 'Congressional Districts': '9', 'WAOFM - GIS - Legislative District Boundary': '49'}
{'_id': ObjectId('67b36f02e23a723cde1e90df'), 'VIN (1-10)': '5YJ3E1EB6K', 'County': 'King', 'City': 'Bothell', 'State': 'WA', 'Postal Code': '98011', 'Model Year': '2019', 'Make': 'TESLA', 'Model': 'MODEL 3', 'Electric Vehicle Type': 'Battery Electric Vehicle (BEV)', 'Clea

In [546]:
# Checking for duplicates of vin in the data- because I am using it as primary key while inserting into SQLlite database
vin_set = set()
duplicates = 0
for record in extracted_data:
    vin = record.get("VIN (1-10)", None)
    if vin in vin_set:
        duplicates += 1
    else:
        vin_set.add(vin)

print(f"Total duplicates found: {duplicates}")

Total duplicates found: 5845


##### There are around 5845 duplicate records that have duplicate values for VIN. As it is an identification number it must be distinct. Therefore I will keep the records unique as I will be using it as a primary key in SQLlite

In [548]:
#code to keep only unique values
unique_data = []
vin_set = set()

for record in extracted_data:
    vin = record.get("VIN (1-10)", None)
    if vin not in vin_set:
        vin_set.add(vin)
        unique_data.append(record)  # Keeping only unique records

print(f"Total unique records: {len(unique_data)}")

Total unique records: 4155


##### Step 4: Connecting to the SQL lite database

In [550]:
import sqlite3 #importing sqllite

conn = sqlite3.connect('ElectricVehiclesData.sqlite') # connecting to created database in SQLlite
cur = conn.cursor() # Initilaizing cursor

# Creating table if not exists
cur.execute("""
        CREATE TABLE IF NOT EXISTS ElectricVehicles (
            VIN TEXT PRIMARY KEY,
            County TEXT,
            City TEXT,
            State TEXT,
            PostalCode TEXT,
            ModelYear INTEGER,
            Make TEXT,
            Model TEXT,
            VehicleType TEXT,
            CAFV_Eligibility TEXT,
            ElectricRange INTEGER,
            BaseMSRP INTEGER,
            LegislativeDistrict TEXT,
            DOLVehicleID INTEGER,
            VehicleLocation TEXT,
            ElectricUtility TEXT,
            CensusTract TEXT,
            Counties TEXT,
            CongressionalDistricts TEXT,
            LegislativeBoundary TEXT
        )
    """)
conn.commit() # committing the action

##### Step 5: Verifying if the table is created successfully

In [552]:
# Verifying if the table is created successfully
cur.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='ElectricVehicles';")
table_check = cur.fetchone()

if table_check:
    print("Table 'ElectricVehicles' created") #Success
else:
    print("Table 'ElectricVehicles' creation failed.")# failure

Table 'ElectricVehicles' created


##### Step 6: Validating data and Inserting data into SQLlite

In [554]:
# Inserting data into SQLite
for record in unique_data:
    try:
        # Extracting and validating each field
        # If any column or key is missing in document then it will return None so that is why get("column name", None) is used for all columns
        vin = record.get("VIN (1-10)", None)
        county = record.get("County", None)
        city = record.get("City", None)
        state = record.get("State", None)
        postal_code = record.get("Postal Code", None)
        model_year = int(record.get("Model Year", 0))  # Converting to integer
        make = record.get("Make", None)
        model = record.get("Model", None)
        vehicle_type = record.get("Electric Vehicle Type", None)
        cafv_eligibility = record.get("Clean Alternative Fuel Vehicle (CAFV) Eligibility", None)
        electric_range = int(record.get("Electric Range", 0))  # Converting to integer
        base_msrp = int(record.get("Base MSRP", 0))  # Converting to integer
        legislative_district = record.get("Legislative District", None)
        dol_vehicle_id = record.get("DOL Vehicle ID", 0)
        vehicle_location = record.get("Vehicle Location", None)
        electric_utility = record.get("Electric Utility", None)
        census_tract = record.get("2020 Census Tract", None)
        counties = record.get("Counties", None)
        congressional_districts = record.get("Congressional Districts", None)
        legislative_boundary = record.get("WAOFM - GIS - Legislative District Boundary", None)

        # Inserting data into SQLite
        # ? is used because the rubric says to consider the parameterized queries while inserting
        cur.execute("""
            INSERT OR REPLACE INTO ElectricVehicles(
                VIN, County, City, State, PostalCode, ModelYear, Make, Model, VehicleType,
                CAFV_Eligibility, ElectricRange, BaseMSRP, LegislativeDistrict, DOLVehicleID,
                VehicleLocation, ElectricUtility, CensusTract, Counties, CongressionalDistricts, LegislativeBoundary
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            vin, county, city, state, postal_code, model_year, make, model, vehicle_type,
            cafv_eligibility, electric_range, base_msrp, legislative_district, dol_vehicle_id,
            vehicle_location, electric_utility, census_tract, counties, congressional_districts, legislative_boundary
        ))

    except Exception as insert_error:
        print("Skipping record due to error:", insert_error)

# Committing all insertions
conn.commit()

# Printing total rows inserted
cur.execute("SELECT COUNT(*) FROM ElectricVehicles")
total_rows = cur.fetchone()[0]
print("Total records inserted into SQLite:", total_rows)  # Success


Total records inserted into SQLite: 4155


##### **NOTE**:
##### The above mentioned total records count in sqllite is accurate because even though I extracted **10,000** records from mongodb at the begining I removed the duplicate values as I used it as primary key in database.  

##### **VIN stands for vehicle identification number**

##### Step 7:  Validating for any null values or empty values in database

In [557]:
# Checking for NULL values in each column
columns = [
    "VIN", "County", "City", "State", "PostalCode", "ModelYear", "Make", "Model", "VehicleType",
    "CAFV_Eligibility", "ElectricRange", "BaseMSRP", "LegislativeDistrict", "DOLVehicleID",
    "VehicleLocation", "ElectricUtility", "CensusTract", "Counties", "CongressionalDistricts", "LegislativeBoundary"
]

# Checking for NULL values in each column
for column in columns:
    query = f"SELECT COUNT(*) FROM ElectricVehicles WHERE {column} IS NULL OR {column} = ''"
    cur.execute(query)
    null_count = cur.fetchone()[0]
    print(f"{column} has {null_count} NULL/empty values.")

VIN has 0 NULL/empty values.
County has 0 NULL/empty values.
City has 0 NULL/empty values.
State has 0 NULL/empty values.
PostalCode has 0 NULL/empty values.
ModelYear has 0 NULL/empty values.
Make has 0 NULL/empty values.
Model has 0 NULL/empty values.
VehicleType has 0 NULL/empty values.
CAFV_Eligibility has 0 NULL/empty values.
ElectricRange has 0 NULL/empty values.
BaseMSRP has 0 NULL/empty values.
LegislativeDistrict has 0 NULL/empty values.
DOLVehicleID has 0 NULL/empty values.
VehicleLocation has 0 NULL/empty values.
ElectricUtility has 0 NULL/empty values.
CensusTract has 0 NULL/empty values.
Counties has 3 NULL/empty values.
CongressionalDistricts has 0 NULL/empty values.
LegislativeBoundary has 0 NULL/empty values.


##### It's great that none of the columns contain null or empty values!

##### Step 8: Printing out the rows from the sqllite table 

In [560]:
from tabulate import tabulate 
cur.execute("SELECT * FROM ElectricVehicles")
all_records = cur.fetchall()

# Fetch column names
column_names = [desc[0] for desc in cur.description]

# Printing records 
print("All records in SQLite:")
print(column_names)  # Printing column headers
for record in all_records:
    print(record)

All records in SQLite:
['VIN', 'County', 'City', 'State', 'PostalCode', 'ModelYear', 'Make', 'Model', 'VehicleType', 'CAFV_Eligibility', 'ElectricRange', 'BaseMSRP', 'LegislativeDistrict', 'DOLVehicleID', 'VehicleLocation', 'ElectricUtility', 'CensusTract', 'Counties', 'CongressionalDistricts', 'LegislativeBoundary']
('2T3YL4DV0E', 'King', 'Bellevue', 'WA', '98005', 2014, 'TOYOTA', 'RAV4', 'Battery Electric Vehicle (BEV)', 'Clean Alternative Fuel Vehicle Eligible', 103, 0, '41', 186450183, 'POINT (-122.1621 47.64441)', 'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)', '53033023604', '3009', '9', '49')
('5YJ3E1EB6K', 'King', 'Bothell', 'WA', '98011', 2019, 'TESLA', 'MODEL 3', 'Battery Electric Vehicle (BEV)', 'Clean Alternative Fuel Vehicle Eligible', 220, 0, '1', 478093654, 'POINT (-122.20563 47.76144)', 'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)', '53033022102', '3009', '1', '1')
('5UX43EU02S', 'Thurston', 'Olympia', 'WA', '98502', 2025, 'BMW', 'X5', 'Plug-in Hybrid Electric Vehicl

In [561]:
# Fetching count of records
cur.execute("SELECT COUNT(*) FROM ElectricVehicles")
record_count = cur.fetchone()[0]

print(f"Total number of records: {record_count}")

conn.close()  # Closing database connection

Total number of records: 4155
